### PR TITLE
Add GPU support

### PR DESCRIPTION
**Pull request by davidmirror-ops**
This PR introduces support for requesting an additional AKS node pool with GPU-powered instances. It was tested with multiple combinations of generic GPU device request, specific accelerator, A100s partitioned and unpartitioned.

### CODE REVIEW

**Comment by davidmirror-ops** [comment]
nvm, I see for a proper integration with the ExtendedResources plugin, at least taints are missing.

**Comment by fiedlerNr9** [comment]
Awesome! Yes it would be cool to include the needed `values.yaml`changes + adding an example how to run an example with flytekit like this:

```python 
@task(requests=Resources( gpu="1"))
def gpu_available() -> bool:
   return torch.cuda.is_available()
```

Also just stumbled upon the [gpu-operator](https://github.com/NVIDIA/gpu-operator), since we dont have that with union deployments. In Union we just deploy the nvidia [k8s-device-plugin](https://github.com/NVIDIA/k8s-device-plugin). Curious about how these 2 compare - from what ive quickly read the device-plugin in included in the gpu operator right?

**Comment by davidmirror-ops** [comment]
> device-plugin is included in the gpu operator right
That's correct, I also learned about it from the NVIDIA TF Modules, where they use it exclusively and could confirm that all [the resources](https://github.com/NVIDIA/k8s-device-plugin/tree/main/deployments/helm/nvidia-device-plugin/templates) typically created by the device-driver plugin are there in the cluster

**Comment by davidmirror-ops** [comment]
I'm not covering unpartitioned A100 requirements mainly because there's still a sizeable body of documentation to complete for the rest of the accelerator use cases.

File: environments/azure/flyte-core/README.md [hunk 1]
```diff
@@ -118,6 +118,100 @@ You have a fully working Flyte environment on Azure.
 
 From this point on, you can continue your learning journey by going through the [Flyte Fundamentals](https://docs.flyte.org/en/latest/flyte_fundamentals/index.html) tutorials.
 
+## Consuming GPU accelerators
+
+To be able to request GPUs on Azure directly from your Flyte tasks, you have two main options. This section covers both and how to use them with the Terraform/OpenTofu modules:
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/README.md]: What exactly you mean with two main options? In this ReadMe you are following up with 1,2,3

File: environments/azure/flyte-core/README.md [hunk 2]
```diff
@@ -118,6 +118,100 @@ You have a fully working Flyte environment on Azure.
...
+image = ImageSpec(
+    base_image= "ghcr.io/flyteorg/flytekit:py3.10-1.10.2",
+     name="pytorch",
+     python_version="3.10",
+     packages=["torch"],
+     builder="envd",
+     registry="<YOUR_CONTAINER_REGISTRY>",
+ )
+# This is the image from the above image spec
+image = "<YOUR_IMAGE_REPO:TAG>"
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/README.md]: Why is this needed? Imagespec is automatically assigning `"<YOUR_IMAGE_REPO:TAG>"` when being passed to `container_image`,  no?

File: environments/azure/flyte-core/README.md [hunk 3]
```diff
@@ -118,6 +118,100 @@ You have a fully working Flyte environment on Azure.
...
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/README.md]: same thing

File: environments/azure/flyte-core/README.md [hunk 4]
```diff
@@ -118,6 +118,100 @@ You have a fully working Flyte environment on Azure.
...
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/README.md]: same things

File: environments/azure/flyte-core/values-aks.yaml [hunk 5]
```diff
@@ -242,7 +242,11 @@ configmap:
           - azure.workload.identity/use: "true"
         default-env-vars:
           - AZURE_STORAGE_ACCOUNT_NAME: "{{ .Values.userSettings.storageAccountName }}"
-        default-pod-template-name: "flyte-template" 
+  # Uncomment this line if you plan to request specific accelerator models
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/values-aks.yaml]: Could we not have `gpu-device-node-label` & `gpu-partition-size-node-label` enabled by default? This way we dont would need to adjust the flyte helm chart on this chapter. In terms of the learning purpose of this chapter we could still link these 2 entries. Thoughts?

File: environments/azure/flyte-core/aks.tf [hunk 6]
```diff
@@ -29,3 +40,74 @@ identity {
...
+node_count = local.gpu_node_pool_count
+enable_auto_scaling = true
+min_count           = local.gpu_node_pool_min_count
+max_count           = local.gpu_node_pool_max_count
+vm_size             = local.gpu_machine_type
+os_disk_size_gb    = local.gpu_node_pool_disk_size
+
+#Only used in case you request specific accelerators and/or partitions. Additional configuration may be required.
+# Learn more: https://docs.flyte.org/en/latest/user_guide/productionizing/configuring_access_to_gpus.html
+node_labels = {
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/aks.tf]: Dont want to overcomplicate things here but could we do something like only setting these labels if `local. accelerator` and `local. partition-size` is not None or something?

File: environments/azure/flyte-core/aks.tf [hunk 7]
```diff
@@ -29,3 +40,74 @@ identity {
...
+  "nvidia.com/gpu.device": "${local.accelerator}",
+  "nvidia.com/gpu.partition-size": "${local.partition_size}"
+}
+#The matching toleration is automatically inserted by flytepropeller when you request a GPU (Requests=Resources(gpu=1))
+node_taints = ["nvidia.com/gpu:NoSchedule"]
```
Comment (fiedlerNr9) [inline environments/azure/flyte-core/aks.tf]: In Union we also apply a node taint regarding the accelerator. I think thats for not scheduling the node in the first place - happy to chat about this

**Review by fiedlerNr9 — CHANGES_REQUESTED** [review]
Awesome stuff as always. Would like to test this myself but I was running into gpu quota issues on our sandbox subscription. Would you destroy your current resources, so the gpu capacity is free for me to try out?

Left some comments & happy to chat about them!

File: environments/azure/flyte-core/values-aks.yaml (hunk 5 above)
Comment (davidmirror-ops) [reply environments/azure/flyte-core/values-aks.yaml]: We could but it would apply a nodeSelector that could interfere with scheduling, especially if users are not planning to request named devices or partitions

File: environments/azure/flyte-core/aks.tf (hunk 6 above)
Comment (davidmirror-ops) [reply environments/azure/flyte-core/aks.tf]: Agree, I'll try to find a good way to do it

**Comment by davidmirror-ops** [comment]
@fiedlerNr9 I added the conditional node labeling and tested multiple combinations, including requesting a partition on an A100, and it works!

File: environments/azure/flyte-core/locals.tf [hunk 8]
```diff
@@ -3,9 +3,9 @@ locals {
   environment        = "deployment"
   tenant             = "mytenant"
   location           = "eastus" #Azure region or zone
-  subscription_id    = "8c8589f3-42da-4083-be83-ff9b12412edd"
-  tenant_id          = "35356303-ece0-4649-85bb-c8a9c67fd341"
-  
+  subscription_id    = "<MY_SUBSCRIPTION_ID>"
+  tenant_id          = "<MY_TENANT_ID>"
```
Comment (ctso) [inline environments/azure/flyte-core/locals.tf]: Should consider wiring these up as variables with no default, that way if a user runs `terraform apply` it will prompt them for these values, or they can pass via a `.tfvars` file.

**Review by ctso — APPROVED** [review]

File: environments/azure/flyte-core/locals.tf (hunk 8 above)
Comment (davidmirror-ops) [reply environments/azure/flyte-core/locals.tf]: That was my biggest existential question about this so far. Totally agree! I'll switch it to tfvars. Thanks
